protocol-designer/src/load-file/migration/utils/getEquipmentLoadInfoFromCommands.ts

```
@@ -0,0 +1,101 @@
+import type { Labware, Modules, Pipettes } from '../../../file-types'
+
+export interface EquipmentLoadInfoFromCommands {
+  pipettes: Pipettes
+  modules: Modules
+  labware: Labware
+}
+export const getEquipmentLoadInfoFromCommands = (
+  commands: CreateCommand[],
+  labwareDefinitions: {
+    [definitionId: string]: LabwareDefinition2
+  }
+): EquipmentLoadInfoFromCommands => {
+  const loadPipetteCommands = commands.filter(
+    (command): command is LoadPipetteCreateCommand =>
+      command.commandType === 'loadPipette'
+  )
+  const loadLabwareCommands = commands.filter(
+    (command): command is LoadLabwareCreateCommand =>
+      command.commandType === 'loadLabware'
+  )
+  const loadModuleCommands = commands.filter(
+    (command): command is LoadModuleCreateCommand =>
+      command.commandType === 'loadModule'
+  )
+
+  const pipettes = loadPipetteCommands.reduce<Pipettes>(
+    (acc, loadPipette: LoadPipetteCreateCommand) => ({
+      ...acc,
+      [loadPipette.params.pipetteId]: {
+        pipetteName: loadPipette.params.pipetteName as PipetteName,
+      },
+    }),
+    {}
+  )
+
+  const labware = loadLabwareCommands.reduce<Labware>(
+    (acc, loadLabware: LoadLabwareCreateCommand) => {
+      const { params } = loadLabware
+      const {
+        displayName: nickName,
+        loadName,
+        version,
+        namespace,
+        labwareId,
+      } = params
+      const labwareDefURI = `${namespace}/${loadName}/${version}`
+
+      if (labwareId == null) {
+        console.error(
+          `expected to find a labwareId from loadLabware command but could not with labwareDefURI ${labwareDefURI}`
+        )
+      }
+      if (labwareDefinitions[labwareDefURI] == null) {
+        console.error(
+          `expected to find matching labware definition with loadname ${loadName} but could not`
+        )
+      }
+      const id = labwareId ?? 'unknown id'
+      const displayName = labwareDefinitions[labwareDefURI].metadata.displayName
```

Is the `if (labwareDefinitions[labwareDefURI] == null)` supposed to guard against something? Wouldn't this line of code still crash if `labwareDefinitions[labwareDefURI]` is in fact null?

---

i guess its to help debug. it shouldn't be null, but if it is, we can see the console errors and know how to fix it faster 😄

---

Oh, I mean like: if you expect the code not to be runnable if `labwareDefinitions[labwareDefURI] == null`, wouldn't it be better to just throw an exception rather than `console.error` and letting the code continue to run?

In the #protocol-designer channel, we often get user reports like "An unknown error has occurred. Contact support with the following message: Cannot read properties of undefined." Our users don't have the Chrome Inspector open, so they won't see any messages in `console.log`, right? If you throw the error message as an exception, at least we'll get better error reports from users :)